rfcs/0155-nixos-migrations.md

```diff
@@ -0,0 +1,118 @@
+
+With the extension of the NixOS Configuration system with migration infrastructure, we can close this conceptual gap and fully realize the value of atomic immutable deployments, as we gain the ability to atomically migrate parts of stateful services along with the packages and definitions.
+
+# Proposed Migration Structure
+
+The following extension to the module system and standard is proposed. Each migration is a Set defined in Nix with the following fields:
+
+- `version`
+- `up`
+- `down`
+
+`up` and `down` are both Nix Sets with the following fields:
+
+- `script`
+- `warn`
+- `backup`
+- `restore`
+
+## Version Information
+
+For migrations to work we must know what version the stateful parts of this service are currently using; this will be stored in the metadata of the systemd service in a standard location to be determined. This is needed for the same reason as in the database migration context: to determine which migrations are needed to achieve the new state.
+
+```Nix
+{
+  config.migrations = [ { version = "2023-04-07"; ... } ];
+}
+```
+
+The version information will be stored as a formatted date string (such that lexigraphical and chronological sort are the same), representing the date at which the migration was merged into Nixpkgs.
+
+## Up Script
+
+A place in the Nix Module system for describing the imperative steps required to migrate from the previous version to the new version. This will be run as a systemd "oneshot" service, to take advantage of the standard architecture.
+
+```Nix
+{
+  up.script = ''
+    ${pkgs.my-service-cli} --run-migration /etc/service-state
+  '';
+}
+```
+
+# Down Script
+
+A place in the Nix Module system for describing the imperative steps required to migrate from the current version to the previous version. This will run as a systemd "oneshot" service, to take advantage of the standard architecture. 
+
+```Nix
+{
+  down.script = "my-service.up" ''
+    ${pkgs.my-service-cli} --run-migration /etc/service-state
+  '';
+}
+```
+
+## Transactionality and Warnings
+
+There are limitations for to the effectiveness of up and down scripts. For example, some migrations are destructive in nature, meaning returning to a previous version is not possible. Sometimes a service cannot be upgraded. Human error and entropy can cause these scripts to fail to execute. In a database migration context we can rely on transactions to cope with some of these issues, a luxury we do not have in this context. As such the onus is placed on the module author to provide these properties to the best of their ability. 
+
+### Known destructive migration
+
+```Nix
+{
+  up.warn = ''
```

Comment:
I don't think a warning is sufficient here. It would be nice if there was "immediately stop and force the user to read the warning and confirm if he accepts the risk" thing in Nixpkgs, but there is none. So I think a better way would be to throw an error outright and have the way to bypass it with something like `forceMigration = true;` in the config.

This is kinda like the thing with unfree software, in a way. We REALLY don't want to do anything potentially problematic, so it's better to be left opt-in.